scripts/test-react.js

```diff
@@ -0,0 +1,179 @@
+let babel = require("babel-core");
+let React = require("react");
+let ReactTestRenderer = require("react-test-renderer");
+let { normalize } = require("../lib/utils/json.js");
+/* eslint-disable no-undef */
+let { expect, describe, it } = global;
+
+let reactTestRoot = path.join(__dirname, "../test/react/");
+let prepackOptions = {
+  compatibility: "react-mocks",
+  errorHandler: diag => "Fail",
+  internalDebug: true,
+  serialize: true,
+  uniqueSuffix: "",
+  maxStackDepth: 100,
+  reactEnabled: true,
+  inlineExpressions: true,
+  omitInvariants: true,
+};
+
+function compileSourceWithPrepack(source) {
+  let code = `(function(){${source}})()`;
+  let serialized = prepackSources([{ filePath: "", fileContents: code, sourceMapContents: "" }], prepackOptions);
+  // add the React require back in, as we've removed it with our Prepack mock
+  let compiledSource = serialized.code.replace(/_\$[\d].React/, "React = require('react')");
+  if (serialized == null || serialized.reactStatistics == null) {
+    throw new Error("React test runner failed during serialization");
+  }
+  return {
+    // replace the code to put back the generator (Prepack doesn't serialize them yet)
+    compiledSource,
+    statistics: serialized.reactStatistics,
+  };
+}
+
+function runSource(source) {
+  let codeAfterBabel = babel.transform(source, {
+    presets: ["babel-preset-react"],
+    plugins: ["transform-object-rest-spread"],
+  }).code;
+  /* eslint-disable no-new-func */
+  let fn = new Function("require", "module", codeAfterBabel);
+  let moduleShim = { exports: null };
+  let requireShim = name => {
+    switch (name) {
+      case "react":
+        return React;
+      default:
+        throw new Error(`Unrecognized import: "${name}".`);
+    }
+  };
+  let global = {
+    require: requireShim,
+    module: moduleShim,
+    Object,
+    String,
+  };
+  try {
+    // $FlowFixMe flow doesn't new Function
+    fn.call(global, requireShim, moduleShim);
+  } catch (e) {
+    console.log(codeAfterBabel);
+    throw e;
+  }
+  return moduleShim.exports;
+}
+
+async function runTest(directory, name) {
+  let source = fs.readFileSync(path.join(reactTestRoot, directory, name)).toString();
+  let { compiledSource } = compileSourceWithPrepack(source);
+
+  let A = runSource(source);
+  expect(typeof A).toBe("function");
+  let B = runSource(compiledSource);
+  expect(typeof B).toBe("function");
+
+  let rendererA = ReactTestRenderer.create(null);
+  let rendererB = ReactTestRenderer.create(null);
+
+  if (A == null || B == null) {
+    throw new Error("React test runner issue");
+  }
+  // // Use the original version of the test in case transforming messes it up.
+  let { getTrials } = A;
+  // // Run tests that assert the rendered output matches.
+  let resultA = getTrials(rendererA, A);
+  let resultB = getTrials(rendererB, B);
+
+  // // the test has returned many values for us to check
+  if (Array.isArray(resultA) && Array.isArray(resultA[0])) {
```

Comment:
Can we restructure this so that instead of repeating similar logic in both branches, we first wrap `resultA` and `resultB` into an array if necessary and then do the assertions unconditionally?